vnl_matrix multiple definitions

### Description

Build fails with shared libraries, tests and CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS.

### Steps to Reproduce

Start CMake from Visual Studio command prompt with following options
`cmake -G Ninja -DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_TESTING:BOOL=ON /path/to/ITK/source`

### Expected behavior

Build completes successfully

### Actual behavior

Link error is generated: `itkvnl-5.0.lib(itkvnl-5.0.dll) : error LNK2005: "public: __cdecl vnl_matrix<double>::vnl_matrix<double>(void)" (??0?$vnl_matrix@N@@QEAA@XZ) already defined in itkCoxDeBoorBSplineKernelFunctionTest.cxx.obj`

### Reproducibility

Always. Here is a link to a dashboard build: https://open.cdash.org/viewBuildError.php?buildid=5660838


### Environment

Windows 10 and VS2015 or VS2017.

--- a/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_matrix.h
+++ b/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_matrix.h
@@ -114,12 +114,7 @@ class VNL_EXPORT vnl_matrix
 {
  public:
   //: Default constructor creates an empty matrix of size 0,0.
-  vnl_matrix() :
-    num_rows(0),
-    num_cols(0),
-    data(nullptr)
-  {
-  }
+  vnl_matrix();
 
   //: Construct a matrix of size r rows by c columns
   // Contents are unspecified.

--- a/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_matrix.hxx
+++ b/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_matrix.hxx
@@ -142,7 +142,6 @@ vnl_matrix< T >::vnl_matrix()
 
 //: Creates a matrix with given number of rows and columns.
 // Elements are not initialized. O(m*n).
-
 template <class T>
 vnl_matrix<T>::vnl_matrix (unsigned rowz, unsigned colz)
 : num_rows(rowz), num_cols(colz)
@@ -152,7 +151,6 @@ vnl_matrix<T>::vnl_matrix (unsigned rowz, unsigned colz)
 }
 
 //: Creates a matrix with given number of rows and columns, and initialize all elements to value. O(m*n).
-
 template <class T>
 vnl_matrix<T>::vnl_matrix (unsigned rowz, unsigned colz, T const& value)
 : num_rows(rowz), num_cols(colz)
@@ -187,7 +185,6 @@ vnl_matrix<T>::vnl_matrix(unsigned r, unsigned c, vnl_matrix_type t)
 
 #if 1 // fsm: who uses this?
 //: Creates a matrix with given dimension (rows, cols) and initialize first n elements, row-wise, to values. O(m*n).
-
 template <class T>
 vnl_matrix<T>::vnl_matrix (unsigned rowz, unsigned colz, unsigned n, T const values[])
 : num_rows(rowz), num_cols(colz)
@@ -202,7 +199,6 @@ vnl_matrix<T>::vnl_matrix (unsigned rowz, unsigned colz, unsigned n, T const val
 
 //: Creates a matrix from a block array of data, stored row-wise.
 // O(m*n).
-
 template <class T>
 vnl_matrix<T>::vnl_matrix (T const* datablck, unsigned rowz, unsigned colz)
 : num_rows(rowz), num_cols(colz)
@@ -215,7 +211,6 @@ vnl_matrix<T>::vnl_matrix (T const* datablck, unsigned rowz, unsigned colz)
 
 //: Creates a new matrix and copies all the elements.
 // O(m*n).
-
 template <class T>
 vnl_matrix<T>::vnl_matrix (vnl_matrix<T> const& from)
 : num_rows(from.num_rows), num_cols(from.num_cols)

--- a/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_matrix.hxx
+++ b/Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_matrix.hxx
@@ -130,6 +130,16 @@ do { \
   } \
 } while (false)
 
+
+//: Default constructor creates an empty matrix of size 0,0.
+template< class T >
+vnl_matrix< T >::vnl_matrix()
+  : num_rows( 0 )
+  , num_cols( 0 )
+  , data( nullptr )
+{
+}
+
 //: Creates a matrix with given number of rows and columns.
 // Elements are not initialized. O(m*n).
 

